File: src/client/src/crypto.js
import { HKDF_INFO } from "./constants.js";

// P-256 (secp256r1) + HKDF-SHA256 + AES-256-GCM
// Private key: PKCS8 DER bytes (WebCrypto's native format)
// Public key: raw uncompressed SEC1 (65 bytes, leading 0x04)

/**
 * @typedef {Object} Keypair
 * @property {Uint8Array} publicKey  Raw uncompressed SEC1 (65 bytes)
 * @property {Uint8Array} privateKey PKCS8 DER
 */

/** AES-256-GCM session bound to a shared secret. Nonce is prepended to ciphertext on encrypt */
export class Session {
	#key;

	/** @param {CryptoKey} cryptoKey */
	constructor(cryptoKey) {
		this.#key = cryptoKey;
	}

	/**
	 * @param {Uint8Array} data
	 * @param {Uint8Array} [aad]
	 * @returns {Promise<Uint8Array>} `nonce(12) || ciphertext || tag(16)`
	 */
	async encrypt(data, aad) {
		const nonce = crypto.getRandomValues(new Uint8Array(12));
		/** @type {AesGcmParams} */
		const params = { name: "AES-GCM", iv: nonce };
		if (aad) params.additionalData = /** @type {BufferSource} */ (aad);
		const ciphertext = await crypto.subtle.encrypt(params, this.#key, /** @type {BufferSource} */ (data));
		const out = new Uint8Array(nonce.length + ciphertext.byteLength);
		out.set(nonce);
		out.set(new Uint8Array(ciphertext), nonce.length);
		return out;
	}

	/**
	 * @param {Uint8Array} bytes `nonce(12) || ciphertext || tag(16)`
	 * @param {Uint8Array} [aad]
	 * @returns {Promise<Uint8Array>}
	 */
	async decrypt(bytes, aad) {
		/** @type {AesGcmParams} */
		const params = { name: "AES-GCM", iv: bytes.slice(0, 12) };
		if (aad) params.additionalData = /** @type {BufferSource} */ (aad);
		const plaintext = await crypto.subtle.decrypt(params, this.#key, /** @type {BufferSource} */ (bytes.slice(12)));
		return new Uint8Array(plaintext);
	}
}

/**
 * Generate a fresh P-256 keypair
 * @returns {Promise<Keypair>}
 */
export async function generateKeypair() {
	const pair = /** @type {CryptoKeyPair} */ (await crypto.subtle.generateKey(
		{ name: "ECDH", namedCurve: "P-256" },
		true,
		["deriveKey", "deriveBits"]
	));
	const publicKey = new Uint8Array(await crypto.subtle.exportKey("raw", pair.publicKey));
	const privateKey = new Uint8Array(await crypto.subtle.exportKey("pkcs8", pair.privateKey));
	return { publicKey, privateKey };
}

/**
 * Derive an AES-GCM `Session` from your private key and the peer's public key
 * @param {Uint8Array} privateKey PKCS8 DER
 * @param {Uint8Array} peerPublicKey Raw uncompressed SEC1 (65 bytes)
 * @returns {Promise<Session>}
 */
export async function deriveSession(privateKey, peerPublicKey) {
	const priv = await crypto.subtle.importKey(
		"pkcs8",
		/** @type {BufferSource} */ (privateKey),
		{ name: "ECDH", namedCurve: "P-256" },
		false,
		["deriveBits"]
	);
	const pub = await crypto.subtle.importKey(
		"raw",
		/** @type {BufferSource} */ (peerPublicKey),
		{ name: "ECDH", namedCurve: "P-256" },
		false,
		[]
	);

	const sharedBits = await crypto.subtle.deriveBits({ name: "ECDH", public: pub }, priv, 256);
	const hkdfKey = await crypto.subtle.importKey(
		"raw",
		/** @type {BufferSource} */ (new Uint8Array(sharedBits)),
		"HKDF",
		false,
		["deriveKey"]
	);

	const aesKey = await crypto.subtle.deriveKey(
		{
			name: "HKDF",
			hash: "SHA-256",
			salt: new Uint8Array(0),
			info: new TextEncoder().encode(HKDF_INFO)
		},
		hkdfKey,
		{ name: "AES-GCM", length: 256 },
		false,
		["encrypt", "decrypt"]
	);

	return new Session(aesKey);
}

/**
 * Compute AES-GCM AAD binding envelope metadata: `SHA256(type|originId|targetId)`
 * @param {string} msgType
 * @param {string} originId
 * @param {string} targetId
 * @returns {Promise<Uint8Array>} 32 bytes
 */
export async function computeAAD(msgType, originId, targetId) {
	const data = new TextEncoder().encode(`${msgType}|${originId}|${targetId}`);
	const hash = await crypto.subtle.digest("SHA-256", data);
	return new Uint8Array(hash);
}

File: src/server/crypto.go
package rootproto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/rand"
	"crypto/sha256"
	"fmt"
	"io"
	"sync"

	"golang.org/x/crypto/hkdf"
)

// HKDF info string shared with the JS client
const hkdfInfo = "root-privacy-encryption"

// Session holds an AES-256-GCM cipher bound to a derived key
type Session struct {
	gcm cipher.AEAD
	mu  sync.Mutex
}

// Keypair holds raw P-256 public and private key bytes
// Public key: 65-byte uncompressed SEC1 (0x04 || X || Y)
// Private key: 32-byte raw scalar
type Keypair struct {
	PublicKey  []byte
	PrivateKey []byte
}

// GenerateKeypair creates a new P-256 keypair
func GenerateKeypair() (*Keypair, error) {
	priv, err := ecdh.P256().GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	return &Keypair{
		PublicKey:  priv.PublicKey().Bytes(),
		PrivateKey: priv.Bytes(),
	}, nil
}

// DeriveSharedSecret performs P-256 ECDH and runs HKDF-SHA256 over the result
// to produce a 32-byte AES key
func DeriveSharedSecret(yourPrivateKey, theirPublicKey []byte) ([]byte, error) {
	curve := ecdh.P256()

	priv, err := curve.NewPrivateKey(yourPrivateKey)
	if err != nil {
		return nil, fmt.Errorf("invalid private key: %w", err)
	}

	pub, err := curve.NewPublicKey(theirPublicKey)
	if err != nil {
		return nil, fmt.Errorf("invalid public key: %w", err)
	}

	secret, err := priv.ECDH(pub)
	if err != nil {
		return nil, fmt.Errorf("ECDH failed: %w", err)
	}

	reader := hkdf.New(sha256.New, secret, nil, []byte(hkdfInfo))
	key := make([]byte, 32)
	if _, err := io.ReadFull(reader, key); err != nil {
		return nil, err
	}
	return key, nil
}

// SessionFromKey builds an AES-GCM session from a 32-byte key
func SessionFromKey(key []byte) (*Session, error) {
	if len(key) != 32 {
		return nil, fmt.Errorf("shared secret must be 32 bytes, got %d", len(key))
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return &Session{gcm: gcm}, nil
}

// Encrypt produces `nonce(12) || ciphertext || tag(16)`
func (s *Session) Encrypt(plaintext, aad []byte) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	nonce := make([]byte, s.gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}
	return s.gcm.Seal(nonce, nonce, plaintext, aad), nil
}

// Decrypt consumes `nonce(12) || ciphertext || tag(16)`
func (s *Session) Decrypt(ciphertext, aad []byte) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	nonceSize := s.gcm.NonceSize()
	if len(ciphertext) < nonceSize {
		return nil, fmt.Errorf("ciphertext too short")
	}
	nonce, body := ciphertext[:nonceSize], ciphertext[nonceSize:]
	return s.gcm.Open(nil, nonce, body, aad)
}

// ComputeAAD binds envelope metadata to the ciphertext: SHA256(type|originId|targetId)
func ComputeAAD(msgType, originID, targetID string) []byte {
	h := sha256.Sum256([]byte(msgType + "|" + originID + "|" + targetID))
	return h[:]
}